File: server/app/models/agent.py
"""
SQLAlchemy ORM — 智能体系统
4 张表: agents, agent_memories, agent_room_assignments, persona_templates
"""

import uuid
from datetime import datetime

from sqlalchemy import (
    TIMESTAMP,
    Boolean,
    CheckConstraint,
    Enum,
    ForeignKey,
    Index,
    Integer,
    SmallInteger,
    String,
    Text,
    UniqueConstraint,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, SoftDeleteMixin, TimestampMixin, UUIDMixin
from shared.constants import AgentLevel, AgentOwnerType, AgentStatus


class Agent(UUIDMixin, TimestampMixin, SoftDeleteMixin, Base):
    __tablename__ = "agents"

    owner_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    owner_type: Mapped[str] = mapped_column(
        Enum(AgentOwnerType, name="agent_owner_type", create_type=False),
        nullable=False,
        server_default="user",
    )
    name: Mapped[str] = mapped_column(String(24), nullable=False)
    avatar_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    bio: Mapped[str | None] = mapped_column(String(500), nullable=True)
    level: Mapped[str] = mapped_column(
        Enum(AgentLevel, name="agent_level", create_type=False),
        nullable=False,
        server_default="L1",
    )
    status: Mapped[str] = mapped_column(
        Enum(AgentStatus, name="agent_status", create_type=False),
        nullable=False,
        server_default="active",
    )
    persona_config: Mapped[dict] = mapped_column(JSONB, nullable=False)
    system_prompt: Mapped[str | None] = mapped_column(Text, nullable=True)
    knowledge_base: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    # 运行时状态
    current_room_id: Mapped[uuid.UUID | None] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("rooms.id"), nullable=True
    )
    is_speaking: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    last_speak_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    speak_count_today: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    task_count_today: Mapped[int] = mapped_column(SmallInteger, nullable=False, server_default="0")

    # 暖场配置
    fallback_timeout_sec: Mapped[int | None] = mapped_column(SmallInteger, server_default="30")
    daily_topic_schedule: Mapped[dict | None] = mapped_column(JSONB, nullable=True)

    memories = relationship("AgentMemory", back_populates="agent", lazy="selectin")

    __table_args__ = (
        CheckConstraint("task_count_today >= 0 AND task_count_today <= 3", name="ck_agents_task_count"),
        Index("idx_agents_owner", "owner_id", "owner_type", postgresql_where=text("deleted_at IS NULL")),
        Index("idx_agents_status", "status"),
        Index("idx_agents_room", "current_room_id", postgresql_where=text("current_room_id IS NOT NULL")),
    )


class AgentMemory(UUIDMixin, Base):
    __tablename__ = "agent_memories"

    agent_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("agents.id"), nullable=False
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
    is_summary: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    token_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    expires_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), server_default=text("NOW()"), nullable=False
    )
    deleted_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)

    agent = relationship("Agent", back_populates="memories")

    __table_args__ = (
        Index("idx_agent_memories_agent", "agent_id", "created_at",
              postgresql_where=text("deleted_at IS NULL")),
    )


class AgentRoomAssignment(UUIDMixin, Base):
    __tablename__ = "agent_room_assignments"

    agent_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("agents.id"), nullable=False
    )
    room_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("rooms.id"), nullable=False
    )
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    assigned_at: Mapped[datetime] = mapped_column(
        TIMESTAMP(timezone=True), server_default=text("NOW()"), nullable=False
    )
    removed_at: Mapped[datetime | None] = mapped_column(TIMESTAMP(timezone=True), nullable=True)

    __table_args__ = (
        UniqueConstraint("agent_id", "room_id", name="uq_agent_room"),
        Index("idx_agent_room_active", "room_id", postgresql_where=text("is_active = TRUE")),
    )


class PersonaTemplate(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "persona_templates"

    name: Mapped[str] = mapped_column(String(24), nullable=False)
    description: Mapped[str | None] = mapped_column(String(200), nullable=True)
    avatar_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    persona_config: Mapped[dict] = mapped_column(JSONB, nullable=False)
    is_onboarding: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    sort_order: Mapped[int] = mapped_column(SmallInteger, nullable=False, server_default="0")

File: app/lib/models/agent.dart
/// AetherVerse — 智能体数据模型
library;

/// 人格配置
class PersonaConfig {
  final String? personality;
  final String? speakingStyle;
  final String? expertise;
  final String? constraints;

  PersonaConfig({this.personality, this.speakingStyle, this.expertise, this.constraints});

  factory PersonaConfig.fromJson(Map<String, dynamic> json) {
    return PersonaConfig(
      personality: json['personality'] as String?,
      speakingStyle: json['speaking_style'] as String?,
      expertise: json['expertise'] as String?,
      constraints: json['constraints'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    if (personality != null) 'personality': personality,
    if (speakingStyle != null) 'speaking_style': speakingStyle,
    if (expertise != null) 'expertise': expertise,
    if (constraints != null) 'constraints': constraints,
  };
}

/// 智能体
class Agent {
  final String id;
  final String ownerId;
  final String ownerType; // user, system
  final String name;
  final String? avatarUrl;
  final String? bio;
  final String level; // L1, L2
  final String status; // active, paused, suspended
  final PersonaConfig? personaConfig;
  final String? currentRoomId;
  final bool isSpeaking;
  final DateTime createdAt;

  Agent({
    required this.id,
    required this.ownerId,
    required this.ownerType,
    required this.name,
    this.avatarUrl,
    this.bio,
    this.level = 'L1',
    this.status = 'active',
    this.personaConfig,
    this.currentRoomId,
    this.isSpeaking = false,
    required this.createdAt,
  });

  factory Agent.fromJson(Map<String, dynamic> json) {
    return Agent(
      id: json['id'] as String,
      ownerId: json['owner_id'] as String,
      ownerType: json['owner_type'] as String? ?? 'user',
      name: json['name'] as String,
      avatarUrl: json['avatar_url'] as String?,
      bio: json['bio'] as String?,
      level: json['level'] as String? ?? 'L1',
      status: json['status'] as String? ?? 'active',
      personaConfig: json['persona_config'] != null
          ? PersonaConfig.fromJson(json['persona_config'] as Map<String, dynamic>)
          : null,
      currentRoomId: json['current_room_id'] as String?,
      isSpeaking: json['is_speaking'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  bool get isActive => status == 'active';
  bool get isPaused => status == 'paused';
  bool get isInRoom => currentRoomId != null;

  /// 复制并修改指定字段
  Agent copyWith({
    String? name,
    String? avatarUrl,
    String? bio,
    String? level,
    String? status,
    PersonaConfig? personaConfig,
    String? currentRoomId,
    bool? isSpeaking,
  }) {
    return Agent(
      id: id,
      ownerId: ownerId,
      ownerType: ownerType,
      name: name ?? this.name,
      avatarUrl: avatarUrl ?? this.avatarUrl,
      bio: bio ?? this.bio,
      level: level ?? this.level,
      status: status ?? this.status,
      personaConfig: personaConfig ?? this.personaConfig,
      currentRoomId: currentRoomId ?? this.currentRoomId,
      isSpeaking: isSpeaking ?? this.isSpeaking,
      createdAt: createdAt,
    );
  }
}

/// 智能体记忆
class AgentMemory {
  final String id;
  final String content;
  final bool isSummary;
  final DateTime createdAt;
  final DateTime? expiresAt;

  AgentMemory({
    required this.id,
    required this.content,
    this.isSummary = false,
    required this.createdAt,
    this.expiresAt,
  });

  factory AgentMemory.fromJson(Map<String, dynamic> json) {
    return AgentMemory(
      id: json['id'] as String,
      content: json['content'] as String,
      isSummary: json['is_summary'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      expiresAt: json['expires_at'] != null
          ? DateTime.parse(json['expires_at'] as String)
          : null,
    );
  }
}

/// 人格模板
class PersonaTemplate {
  final String id;
  final String name;
  final String? description;
  final String? avatarUrl;
  final PersonaConfig? personaConfig;

  PersonaTemplate({
    required this.id,
    required this.name,
    this.description,
    this.avatarUrl,
    this.personaConfig,
  });

  factory PersonaTemplate.fromJson(Map<String, dynamic> json) {
    return PersonaTemplate(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      avatarUrl: json['avatar_url'] as String?,
      personaConfig: json['persona_config'] != null
          ? PersonaConfig.fromJson(json['persona_config'] as Map<String, dynamic>)
          : null,
    );
  }
}
